gratka/utils.py

```diff
@@ -0,0 +1,159 @@
+    """
+    for i, j in dic.items():
+        text = text.replace(i, j)
+    return text
+
+
+def normalize_text(text, lower=True, replace_spaces='_'):
+    """
+    This method returns the input string, but normalizes is it for use in the url.
+    :param text: input string
+    :rtype: string
+    :return: Normalized string. lowercase, no diacritics, '-' instead of ' '
+    """
+    try:
+        unicoded = unicode(text, 'utf8')
+    except TypeError:
+        unicoded = text
+    if lower:
+        unicoded = unicoded.lower()
+    normalized = unicodedata.normalize('NFKD', unicoded)
+    encoded_ascii = normalized.encode('ascii', 'ignore')
+    decoded_utf8 = encoded_ascii.decode("utf-8")
+    if replace_spaces:
+        decoded_utf8 = decoded_utf8.replace(" ", replace_spaces)
+    return decoded_utf8
+
+
+def get_region_from_autosuggest(region_part):
+    """
+    This method makes a request to the OtoDom api, asking for the best fitting region for the supplied region_part string.
+    :param region_part: input string, it should be a part of an existing region in Poland, either city, street, district or voivodeship
+    :rtype: dict
+    :return: A dictionary which contents depend on the API response.
+    """
+    if not region_part:
+        return {}
+    url = u"http://www.gratka.pl/b-dom/ajax/podpowiedzi-lokalizacja/?tekst={0}".format(region_part)
+    response = json.loads(get_response_for_url(url).text)[0]
+
+    region_dict = {}
+
+    if "powiat" in response:
+        region_dict["county"] = normalize_text(response["powiat"])
+    if "miejscowosc" in response:
+        region_dict["city"] = normalize_text(response["miejscowosc"])
+    if "ulica" in response:
+        region_dict["street"] = normalize_text(response["ulica"])
+    if "dzielnica" in response:
+        region_dict["district"] = normalize_text(response["dzielnica"])
+
+    return region_dict
+
+
+def get_url(main_category, detail_category, voivodeship, region, page=1, **filters):
+    """
+    This method builds a ready-to-use url based on the input parameters.
+    :param main_category: see :meth:`gratka.category.get_category` for reference
+    :param detail_category: see :meth:`gratka.category.get_category` for reference
+    :param voivodeship: see :meth:`gratka.category.get_category` for reference
+    :param region: see :meth:`gratka.category.get_category` for reference
+    :param page: page number
+    :param filters: see :meth:`gratka.category.get_category` for reference
+    :rtype: string
+    :return: the url
+    """
+    if main_category != "inwestycje" and main_category != "pokoje-do-wynajecia":
+        detail_category = "-".join([main_category, detail_category])
+    else:
+        detail_category = main_category
+    region_dict = get_region_from_autosuggest(region)
+    basic_url = "/".join([BASE_URL, detail_category, "lista"]) + "/" + ",".join([voivodeship, region_dict.get("city", "")])
+    if "keywords" in filters:
```

Comment:
maybe write comment that here start filter section or move it into some separate function (in that case it might be easier to test)